src/nvim/diff.c

```diff
@@ -1733,25 +1743,713 @@ void diff_clear(tabpage_T *tp)
+      initialize_compareline3(dp, b1, p1, b0, b2);
+      initialize_compareline3(dp, b2, p2, b0, b1);
+    } else if (p == DFPATH3_COMPARE012) {
+      dp->df_comparisonlines[dp->df_arr_col_size * b0 + p0].compare[b1] = p1;
+      dp->df_comparisonlines[dp->df_arr_col_size * b0 + p0].compare[b2] = p2;
+      dp->df_comparisonlines[dp->df_arr_col_size * b1 + p1].compare[b0] = p0;
+      dp->df_comparisonlines[dp->df_arr_col_size * b1 + p1].compare[b2] = p2;
+      dp->df_comparisonlines[dp->df_arr_col_size * b2 + p2].compare[b0] = p0;
+      dp->df_comparisonlines[dp->df_arr_col_size * b2 + p2].compare[b1] = p1;
+      p0++, p1++, p2++;
+      initialize_compareline3(dp, b0, p0, b1, b2);
+      initialize_compareline3(dp, b1, p1, b0, b2);
+      initialize_compareline3(dp, b2, p2, b0, b1);
+    } else if  (p == DFPATH3_SKIP0) {
+      dp->df_comparisonlines[dp->df_arr_col_size * b2 + p2].filler++;
+      dp->df_comparisonlines[dp->df_arr_col_size * b1 + p1].filler++;
+      dp->df_comparisonlines[dp->df_arr_col_size * b0 + p0].newline = true;
+      p0++;
+      initialize_compareline3(dp, b0, p0, b1, b2);
+    } else if (p == DFPATH3_SKIP1) {
+      dp->df_comparisonlines[dp->df_arr_col_size * b0 + p0].filler++;
+      dp->df_comparisonlines[dp->df_arr_col_size * b2 + p2].filler++;
+      dp->df_comparisonlines[dp->df_arr_col_size * b1 + p1].newline = true;
+      p1++;
+      initialize_compareline3(dp, b1, p1, b0, b2);
+    } else if (p == DFPATH3_SKIP2) {
+      dp->df_comparisonlines[dp->df_arr_col_size * b0 + p0].filler++;
+      dp->df_comparisonlines[dp->df_arr_col_size * b1 + p1].filler++;
+      dp->df_comparisonlines[dp->df_arr_col_size * b2 + p2].newline = true;
+      p2++;
+      initialize_compareline3(dp, b2, p2, b0, b1);
+    }
+  }
+  for (int i = 0; i < (2); i++) {
+    for (int j = 0; j < (dp->df_count[b1]+1); j++) {
+      for (int k = 0; k < (dp->df_count[b2]+1); k++) {
+        xfree(df_pathmatrix3[i][j][k].df_path3);
+      }
+      xfree(df_pathmatrix3[i][j]);
+    }
+    xfree(df_pathmatrix3[i]);
+  }
+  xfree(df_pathmatrix3);
+  xfree(mem12), xfree(mem01), xfree(mem02);
+}
+
+
 /// Check diff status for line "lnum" in buffer "buf":
 ///
-/// Returns 0 for nothing special
-/// Returns -1 for a line that should be highlighted as changed.
-/// Returns -2 for a line that should be highlighted as added/deleted.
-/// Returns > 0 for inserting that many filler lines above it (never happens
-/// when 'diffopt' doesn't contain "filler").
-/// This should only be used for windows where 'diff' is set.
+/// Returns 0 for nothing special Returns -1 for a line that should be
+/// highlighted as changed.  Returns -2 for a line that should be
+/// highlighted as added/deleted.  Returns > 0 for inserting that many
+/// filler lines above it (never happens when 'diffopt' doesn't contain
+///     "filler").  This should only be used for windows where 'diff' is
+/// set.  When diffopt contains linematch, a changed/added/deleted line
```

Comment:
You broke the formatting here and made it harder to read. I'm guessing this was accidental using `gq`?

Reply:
thanks for pointing that out, fixed